### PR TITLE
Add devise

### PR DESCRIPTION
**Pull request by samuelmoiwa**
## **In this Branch, @KareemWilson  and I did the following:** 

**1. Install and setup devise**
**2. Add the following features:**  :arrow_down: 
- Can register a new user. :heavy_check_mark: 
Capture 3

- User logs in with a combination of email and password. :heavy_check_mark: 
Capture 4

- Hashed password stored in the database. :heavy_check_mark: 

- Ask for confirmation by email. :heavy_check_mark: 

- Can reset password. :heavy_check_mark: 
Capture 5

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,12 @@
 class ApplicationController < ActionController::Base
-  def current_user
-    User.first
+  protect_from_forgery with: :exception
+
+  before_action :update_allowed_parameters, if: :devise_controller?
+
```
Comment (DammyShittu) [inline app/controllers/application_controller.rb]: - [ ] Users who are not signed can still visit different routes in your app. Kindly add the `authenticate_user!` helper to redirect the user to log in if they are not signed in.

  ```
  before_action :authenticate_user!
  ```

File: app/views/devise/sessions/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      <%= f.label :remember_me %>
+    </div>
+  <% end %>
+
+  <div class="actions">
+    <%= f.submit "Log in" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (DammyShittu) [inline app/views/devise/sessions/new.html.erb:26]: - [ ] Kindly modify your form to match the wireframe shared with you. Your form elements are aligned to the left whereas in the wireframe they are centered. Kindly fix that.

  image

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Sign up" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (DammyShittu) [inline app/views/devise/registrations/new.html.erb]: - [ ] Kindly center the sign up form as well

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @samuelmoiwa and @KareemWilson,

## Highlights
✔️ Descriptive PR
✔️ Passing linters
✔️ Good branch name

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
[# STATUS: APPROVED :green_circle: :clap: 

Hello @samuelmoiwa  , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: aaa
Congratulations! 🎉 

HIGHLIGHTS:

- [x]  Your repo is professional 
- [x] Descriptive P.R
- [x] Proffessional README









 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]

**Pull request merged**